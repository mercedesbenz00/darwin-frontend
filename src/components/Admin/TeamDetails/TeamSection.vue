<template>
  <div
    class="section"
    :class="{'section--inactive': !active}"
  >
    <div class="section__title">
      <slot name="title" />
    </div>
    <div class="section__body">
      <slot name="body" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({
  name: 'team-section'
})
export default class TeamSection extends Vue {
  @Prop({ required: false, default: true, type: Boolean as () => boolean })
  active!: boolean
}
</script>

<style lang="scss" scoped>
.section {
  @include row;
  flex-wrap: wrap;
  margin-bottom: 25px;
  padding: 20px;
  background: $colorSecondaryLight3;
  border-radius: 3px;

}

.section .section--inactive {
  opacity: .3;
}

.section .section__title {
  width: 80%;
  margin-right: 20%;
  @include typography(xl, headlines, bold);
  margin-bottom: 10px;
}

.section .section__body {
  @include row;
  flex-wrap: wrap;

  gap: 15px;
}
</style>
