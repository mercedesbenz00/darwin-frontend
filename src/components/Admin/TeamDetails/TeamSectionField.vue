<template>
  <div
    class="field"
    :class="{ 'field--full': fullWidth }"
  >
    <label class="field__label">
      <slot name="label" />
    </label>
    <div class="field__value">
      <slot name="value" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({
  name: 'team-section-field'
})
export default class TeamSectionField extends Vue {
  @Prop({ required: false, default: false, type: Boolean as () => boolean })
  fullWidth!: boolean
}
</script>

<style lang="scss" scoped>
.field {
  width: 300px;
  @include col;
}

.field.field--full {
  width: 100%;
}

.field.field__label {
  font-weight: bold;
  margin-bottom: 10px;
}
</style>
