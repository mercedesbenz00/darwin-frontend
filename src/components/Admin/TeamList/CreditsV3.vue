<template>
  <div class="credits">
    <v-popover trigger="hover">
      <div
        class="credits__info"
        :class="{'credits__info--critical': closeToExpiry || fewCreditsLeft}"
      >
        {{ remainingThisPeriodFormatted }}
      </div>
      <template slot="popover">
        <div class="credits__tooltip-row">
          <label>Billed this period</label>
          <span>{{ billedThisPeriodFormatted }}</span>
        </div>
        <div class="credits__tooltip-row">
          <label>Remaining this period</label>
          <span>{{ remainingThisPeriodFormatted }}</span>
        </div>
        <div class="credits__tooltip-row">
          <label>Billed next period</label>
          <span>{{ billedNextPeriodFormatted }}</span>
        </div>
        <div
          v-if="daysToNextExpiry"
          class="credits__tooltip-row"
        >
          <label>Next credits expire in {{ daysToNextExpiry }} day(s).</label>
        </div>
        <div
          v-else-if="remainingThisPeriod < 100"
          class="credits__tooltip-row"
        >
          <label>This team has less than 100 hours of annotation time left.</label>
        </div>
        <div
          v-else
          class="credits__tooltip-row"
        >
          This team has no hours assigned.
        </div>
      </template>
    </v-popover>

    <add-credits :team="team" />
  </div>
</template>
<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import { TeamPayload } from '@/store/modules/admin/types'
import { CustomerSubscriptionPayload } from '@/store/modules/billing/types'
import { dateDiff, dateUtcNow, dateFromUtcISO } from '@/utils'
import {
  adaptAdminTeamPayload,
  annotationCreditsBilledNextPeriod,
  annotationCreditsRemainingThisPeriod,
  annotationCreditsBilledThisPeriod
} from '@/utils/billing'

import AddCredits from './AddCredits.vue'

// `undefined` means "use default client locale"
const format = (value: number): string =>
  value.toLocaleString(undefined, { maximumFractionDigits: 0 })

@Component({ name: 'credit-v3', components: { AddCredits } })
export default class CreditV3 extends Vue {
  @Prop({ required: true })
  team!: TeamPayload

  // info UI

  get subscription (): CustomerSubscriptionPayload {
    if (!this.team.customer_v3) { throw new Error('BillingV3 team missing customer information') }
    return this.team.customer_v3.customer_subscription
  }

  get adaptedInfo (): ReturnType<typeof adaptAdminTeamPayload> {
    return adaptAdminTeamPayload(this.team)
  }

  get billedThisPeriod (): number {
    return annotationCreditsBilledThisPeriod(this.adaptedInfo)
  }

  get billedThisPeriodFormatted (): string {
    return format(this.billedThisPeriod)
  }

  get remainingThisPeriod (): number {
    return annotationCreditsRemainingThisPeriod(this.adaptedInfo)
  }

  get remainingThisPeriodFormatted (): string {
    return format(this.remainingThisPeriod)
  }

  get billedNextPeriod (): number {
    return annotationCreditsBilledNextPeriod(this.adaptedInfo)
  }

  get billedNextPeriodFormatted (): string {
    return format(this.billedNextPeriod)
  }

  get daysToNextExpiry (): number | null {
    if (!this.team.credit_next_expiry) { return null }
    const nextExpiry = dateFromUtcISO(this.team.credit_next_expiry)
    const now = dateUtcNow()
    const diff = dateDiff(nextExpiry, now, 'days')
    return Math.ceil(diff)
  }

  get closeToExpiry (): boolean {
    return !!this.daysToNextExpiry && this.daysToNextExpiry <= 7
  }

  get fewCreditsLeft (): boolean {
    return this.remainingThisPeriod < 100
  }
}
</script>
<style lang="scss" scoped>
.credits {
  @include row--right;
  width: 100%;
}

.credits__info {
  margin-right: 10px;
}

.credits__info--critical {
  font-weight: bold;
  color: $colorPink;
}

.credits__tooltip-row {
  min-width: 150px;
  margin: 10px;
  @include row--distributed;
}

.credits__tooltip-row label {
  font-weight: bold;
}
</style>
