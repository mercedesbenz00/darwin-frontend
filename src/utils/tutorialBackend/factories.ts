import {
  AnnotateStagePayload,
  CodeStagePayload,
  CompleteStagePayload,
  DatasetImagePayload,
  DatasetItemPayload,
  DatasetItemStatus,
  DatasetItemType,
  ImagePayload,
  ModelStagePayload,
  ReviewStagePayload,
  StageType,
  WorkflowStagePayload,
  WorkflowStageTemplatePayload,
  WorkflowTemplatePayload
} from '@/store/types'

import TutorialError from './TutorialError'
import { dataset } from './data/dataset'

const itemId = (datasetImageId: number) => datasetImageId * 10
const workflowId = (datasetImageId: number) => datasetImageId * 100

const stageId = (datasetImageId: number, stageNumber: number) =>
  workflowId(datasetImageId) + stageNumber

const initializeDatasetImage = (image: ImagePayload): DatasetImagePayload => ({
  dataset_id: dataset.id,
  dataset_video_id: null,
  id: image.id * 10,
  image,
  seq: image.id,
  set: 1
})

const initializeItem = (datasetImage: DatasetImagePayload): DatasetItemPayload => ({
  archived_reason: null,
  archived: false,
  current_workflow: null,
  current_workflow_id: null,
  dataset_id: datasetImage.dataset_id,
  dataset_image_id: datasetImage.id,
  dataset_image: datasetImage,
  dataset_video_id: null,
  dataset_video: null,
  file_size: 0,
  filename: datasetImage.image.original_filename,
  height: datasetImage.image.height,
  id: itemId(datasetImage.id),
  inserted_at: '2020-06-12T00:00:00',
  labels: [],
  path: '/',
  priority: 0,
  seq: datasetImage.seq,
  set: datasetImage.set,
  status: DatasetItemStatus.annotate,
  updated_at: '2020-06-12T00:00:00',
  width: datasetImage.image.width,
  type: DatasetItemType.image
})

const initializeStage = (
  datasetImage: DatasetImagePayload,
  stageTemplate: WorkflowStageTemplatePayload
): WorkflowStagePayload => {
  const { type } = stageTemplate
  const payload = {
    assignee_id: null,
    completed: false,
    completes_at: null,
    dataset_item_id: itemId(datasetImage.id),
    id: stageId(datasetImage.id, stageTemplate.stage_number),
    metadata: {},
    number: stageTemplate.stage_number,
    skipped_reason: null,
    skipped: false,
    template_metadata: stageTemplate.metadata,
    type: stageTemplate.type,
    workflow_id: workflowId(datasetImage.id),
    workflow_stage_template_id: stageTemplate.id
  }

  if (type === StageType.Annotate) { return payload as AnnotateStagePayload }
  if (type === StageType.Code) { return payload as CodeStagePayload }
  if (type === StageType.Complete) { return payload as CompleteStagePayload }
  if (type === StageType.Model) { return payload as ModelStagePayload }
  if (type === StageType.Review) { return payload as ReviewStagePayload }

  throw new Error('Invalid type given')
}

const initializeWorkflow = (
  datasetImage: DatasetImagePayload,
  workflowTemplate: WorkflowTemplatePayload,
  currentStage: number
): Pick<DatasetItemPayload, 'current_workflow' | 'current_workflow_id'> => {
  const annotateTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Annotate)

  const reviewTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Review)

  const completeTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Complete)

  if (!annotateTemplate || !reviewTemplate || !completeTemplate) {
    throw new TutorialError('Initializing workflow using invalid template ')
  }

  const id = workflowId(datasetImage.id)

  const workflow: DatasetItemPayload['current_workflow'] = {
    dataset_item_id: itemId(datasetImage.id),
    id,
    workflow_template_id: workflowTemplate.id,
    current_stage_number: currentStage,
    current_workflow_stage_template_id: annotateTemplate.id,
    status: DatasetItemStatus.annotate,
    stages: {
      1: [initializeStage(datasetImage, annotateTemplate)],
      2: [initializeStage(datasetImage, reviewTemplate)],
      3: [initializeStage(datasetImage, completeTemplate)]
    }
  }

  return { current_workflow: workflow, current_workflow_id: id }
}

export const createItem = (
  image: ImagePayload,
  workflowTeplate: WorkflowTemplatePayload,
  stageNumber: number = 1
): DatasetItemPayload => {
  const datasetImage = initializeDatasetImage(image)
  return {
    ...initializeItem(datasetImage),
    ...initializeWorkflow(datasetImage, workflowTeplate, stageNumber)
  }
}
