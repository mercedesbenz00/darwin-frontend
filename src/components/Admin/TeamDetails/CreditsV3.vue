<template>
  <team-section
    class="credits-v3"
    :active="active"
  >
    <template #title>
      Credits (V3)
    </template>
    <template #body>
      <team-section-field>
        <template #label>
          Billed this period
        </template>
        <template #value>
          {{ format(billedThisPeriod) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Billed next period
        </template>
        <template #value>
          {{ format(billedNextPeriod) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Bonus
        </template>
        <template #value>
          {{ format(bonus) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Used
        </template>
        <template #value>
          {{ format(used) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Remaining this period
        </template>
        <template #value>
          {{ format(remainingThisPeriod) }}
        </template>
      </team-section-field>
      <team-section-field>
        <template #label>
          Seconds per automation action
        </template>
        <template #value>
          <seconds-per-automation-action
            class="credits-v3__field__value"
            :team="team"
          />
        </template>
      </team-section-field>
    </template>
  </team-section>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

import { TeamPayload } from '@/store/modules/admin/types'
import { CustomerSubscriptionPayload } from '@/store/modules/billing/types'
import {
  adaptAdminTeamPayload,
  annotationCreditsBilledNextPeriod,
  annotationCreditsBilledThisPeriod,
  annotationCreditsRemainingThisPeriod
} from '@/utils/billing'

import SecondsPerAutomationAction from './SecondsPerAutomationAction.vue'
import TeamSection from './TeamSection.vue'
import TeamSectionField from './TeamSectionField.vue'

@Component({
  name: 'credits-v3',
  components: { SecondsPerAutomationAction, TeamSection, TeamSectionField }
})
export default class CreditsV3 extends Vue {
  @Prop({ required: true, type: Object as () => TeamPayload })
  team!: TeamPayload

  get active (): boolean {
    return !!this.team.customer_v3
  }

  get subscription (): CustomerSubscriptionPayload | null {
    return this.team.customer_v3 ? this.team.customer_v3.customer_subscription : null
  }

  get adaptedInfo (): ReturnType<typeof adaptAdminTeamPayload> {
    return adaptAdminTeamPayload(this.team)
  }

  get billedThisPeriod (): number {
    return annotationCreditsBilledThisPeriod(this.adaptedInfo)
  }

  get billedNextPeriod (): number {
    return annotationCreditsBilledNextPeriod(this.adaptedInfo)
  }

  get bonus (): number {
    const { subscription } = this
    return subscription ? subscription.annotation_credits_bonus : 0
  }

  get used (): number {
    const { subscription } = this
    return subscription ? subscription.annotation_credits_used : 0
  }

  get remainingThisPeriod (): number {
    return annotationCreditsRemainingThisPeriod(this.adaptedInfo)
  }

  format (value: number): string {
    return value.toLocaleString(undefined, { maximumFractionDigits: 3 })
  }
}
</script>
