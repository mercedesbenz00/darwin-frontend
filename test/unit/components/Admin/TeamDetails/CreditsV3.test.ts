import { shallowMount, createLocalVue } from '@vue/test-utils'

import { buildCustomerSubscriptionPayload } from 'test/unit/factories'
import { buildAdminTeamPayload } from 'test/unit/factories/buildAdminTeamPayload'

import CreditsV3 from '@/components/Admin/TeamDetails/CreditsV3.vue'
import TeamSection from '@/components/Admin/TeamDetails/TeamSection.vue'
import TeamSectionField from '@/components/Admin/TeamDetails/TeamSectionField.vue'

const localVue = createLocalVue()

const periodStart = 1000

const v7 = buildAdminTeamPayload({
  name: 'V7',
  subscription_period_start: periodStart,
  customer_v3: {
    customer_subscription: buildCustomerSubscriptionPayload({
      annotation_credits_bonus: 40,
      annotation_credits_standard: 100,
      annotation_credits_standard_max_in_period: 200,
      annotation_credits_standard_max_in_period_at: periodStart,
      annotation_credits_used: 50,
      seconds_per_automation_action: 22
    })
  }
})

it('matches snapshot', () => {
  const propsData = { team: v7 }
  const stubs = { TeamSection, TeamSectionField }
  const wrapper = shallowMount(CreditsV3, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})
