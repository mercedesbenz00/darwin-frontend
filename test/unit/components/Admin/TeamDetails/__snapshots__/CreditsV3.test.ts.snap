// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="section credits-v3">
  <div class="section__title">
    Credits (V3)
  </div>
  <div class="section__body">
    <div class="field"><label class="field__label">
        Billed this period
      </label>
      <div class="field__value">
        200
      </div>
    </div>
    <div class="field"><label class="field__label">
        Billed next period
      </label>
      <div class="field__value">
        100
      </div>
    </div>
    <div class="field"><label class="field__label">
        Bonus
      </label>
      <div class="field__value">
        40
      </div>
    </div>
    <div class="field"><label class="field__label">
        Used
      </label>
      <div class="field__value">
        50
      </div>
    </div>
    <div class="field"><label class="field__label">
        Remaining this period
      </label>
      <div class="field__value">
        190
      </div>
    </div>
    <div class="field"><label class="field__label">
        Seconds per automation action
      </label>
      <div class="field__value">
        <seconds-per-automation-action-stub team="[object Object]" class="credits-v3__field__value"></seconds-per-automation-action-stub>
      </div>
    </div>
  </div>
</div>
`;
